### PR TITLE
Added Bean Post-Processor (BPP) 

### PR DESCRIPTION
**Pull request by ArSiu**
for creating proxies of beans requiring user notifications

### CODE REVIEW

**Comment by ajax-gnatiuk-s** [comment]
@YuliaShvets please review

File: src/main/kotlin/com/arsiu/eduhub/EduHubApplication.kt [hunk 1]
```diff
@@ -2,11 +2,13 @@ package com.arsiu.eduhub
 
 import org.springframework.boot.autoconfigure.SpringBootApplication
 import org.springframework.boot.runApplication
+import org.springframework.scheduling.annotation.EnableAsync
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/arsiu/eduhub/EduHubApplication.kt]: unused import?

File: src/main/kotlin/com/arsiu/eduhub/EduHubApplication.kt [hunk 2]
```diff
@@ -2,11 +2,13 @@ package com.arsiu.eduhub
...
 
 @SpringBootApplication
 class EduHubApplication
 
 @Suppress("SpreadOperator")
+
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/arsiu/eduhub/EduHubApplication.kt]: remove blank line

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+                    if (annotation != null) {
+                        GlobalScope.launch {
+                            val info = when (result) {
+                                is Course -> result.name
+                                is Chapter -> result.course.name
+                                is Lesson -> result.chapter.course.name
+                                is Assignment -> result.lesson.chapter.course.name
+                                else -> ""
+                            }
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: BBP is a part of Spring infrastructure. BBP must be flexible and reusable(suppose you'd like someday extract library and add it to another project).

Moreover, when you add some new entity then you must update this BPP as well to support the new entity. 
Sounds like you have strong coupling here

```kotlin
                            val info = when (result) {
                                is Course -> result.name
                                is Chapter -> result.course.name
                                is Lesson -> result.chapter.course.name
                                is Assignment -> result.lesson.chapter.course.name
                                else -> ""
                            }
```

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                        }
+                    }
+
+                    return@InvocationHandler result
+                }
+            )
+        }
+
+        return bean
+    }
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: Have you run detekt?

I believe it reports high code complexity on this function. What about splitting to few smaller and simpler functions?

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    fun notifyUsers(msg: String) {
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: I believe it should be private

File: src/main/kotlin/com/arsiu/eduhub/service/util/mail/EmailServiceImpl.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    private val jms: JavaMailSender,
+    @Value("\${spring.mail.username}") private val sender: String
+) : EmailService {
+
+    override fun sendMail(details: EmailParameters) {
+        val smm = SimpleMailMessage()
+        smm.from = sender
+        smm.setTo(details.receiver)
+        smm.text = details.msg
+        smm.subject = details.subject
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/arsiu/eduhub/service/util/mail/EmailServiceImpl.kt]: It's time to get familiar with [apply](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/apply.html) function

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt (hunk 4 above)
Comment (ArSiu) [reply src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: detekt report
Before
image
After
image

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+) : BeanPostProcessor {
+
+    val map = mutableMapOf<String, KClass<*>>()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = bean::class
+
+        for (method in beanClass.java.methods) {
+            val annotation = method.getAnnotation(NotifyOnChange::class.java)
+            if (annotation != null) {
```
Comment (YuliaShvets) [inline src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: Maybe you can simplify it with let function

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        return bean
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = map[beanName]
+
+        if (beanClass != null) {
```
Comment (YuliaShvets) [inline src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: and here too :3

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                beanClass.java.interfaces,
+                InvocationHandler { _, method, args ->
+                    val result = method.invoke(bean, *(args ?: emptyArray()))
+
+                    val annotation = beanClass.memberFunctions.firstOrNull {
+                        (it.name == method.name) &&
+                                (it::class.java.typeParameters.contentEquals(method::class.java.typeParameters))
+                    }?.findAnnotation<NotifyOnChange>()
+
+                    if (annotation != null) {
```
Comment (YuliaShvets) [inline src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: and here :)

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyTriggerAnnotationHandlerBeanPostProcessor.kt [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            (it.javaClass.typeParameters.contentEquals(method.javaClass.typeParameters))
+        }?.findAnnotation<NotifyTrigger>()
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    private fun notifyUsers(msg: String) {
+        GlobalScope.launch {
+            userService.findAll()
+                .filter { it.role == Role.STUDENT }
+                .forEach { emailServiceImpl.sendMail(EmailParameters(it.email, msg, msg)) }
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/arsiu/eduhub/bpp/NotifyTriggerAnnotationHandlerBeanPostProcessor.kt:65]: I don't ask to do any changes here.

Here you find all the users and sequentially, one by one, send email. 
It's task for you to think about it and later we will discusse

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Review by YuliaShvets — APPROVED** [review]

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt (hunk 7 above)
Comment (ArSiu) [reply src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: done

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt (hunk 9 above)
Comment (ArSiu) [reply src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: done

File: src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt (hunk 8 above)
Comment (ArSiu) [reply src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt]: done

**Pull request merged**